OKTA-594692 policy simulation api documentation (#4105)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/reference/api/policy/index.md

HTTP 200:
Array of [Application objects](/docs/reference/api/apps/#application-object)

> ## Policy Simulation operations

Reviewer: ```suggestion
## Policy simulation operations
```

Reviewer: Our heading style is sentence case

<ApiLifecycle access="ie" />

> > **Note:** This feature is only available as a part of the Identity Engine. Please [contact support](mailto:dev-inquiries@okta.com) for further information.

Reviewer 2: I think you need a line break before `> **Note:**` for it to render properly

> ### Access Simulation

Reviewer: ```suggestion
### Access simulation
```

> Access simulation API is an admin API that evaluates policy and policy rules based on the existing policy rule configuration. The evaluation result simulates what the real world authentication flow is and what policy rules have been applied or matched to the authentication flow.

Reviewer: ```suggestion
The access simulation API is an admin API that evaluates policy and policy rules based on the existing policy rule configuration. The evaluation result simulates what the real world authentication flow is and what policy rules have been applied or matched to the authentication flow.
```

<ApiOperation method="post" url="/api/v1/policies/simulate" />

#### Request
The section below explains the request parameters and the properties inside the request body.

##### Request parameters
| Parameter  | Type   | Description                                       |
| ---------- | ------ | ------------------------------------------------- |

> | `expand` | String | (Optional) `expand=EVALUATED` to include list of evaluated but not matched policy and policy rules `expand=RULE` include detail about why a rule condition (not)matched|

Reviewer: ```suggestion
| `expand` | String | (Optional) Use `expand=EVALUATED` to include a list of evaluated but not matched policy and policy rules. Use `expand=RULE` to include details about why a rule condition was (not) matched. |
```

##### Request Body
| Property | Type                     | Description                          |
| -------- | -------------------------|--------------------------------------|

> | `policyTypes` | Array| (Optional) Supported PolicyTypes for simulation: OKTA_SIGN_ON, MFA_ENROLL, PROFILE_ENROLLMENT, ACCESS_POLICY. Default `null`, return all types|

Author: yes that is correct, default just not supply this property and it will evaluate all supported policy types

| `appInstance` | String | (Required) The appInstance ID for this simulation |

> | `policyContext.user.id` | String | (Required) The userId for this simulation. Only userId or groupIds allowed, not both  |

Reviewer: would it be better to expand `userId` to user ID? Similar comment to `groupIds`?

| `policyContext.groups.ids` | Array | (Required) The groupIds for this simulation. Only userId or groupIds allowed, not both |

> | `policyContext.ip` | String | (Optional) The network rule condition, zone or IP address. See [Network Condition](#network-condition-object) |

Reviewer: ```suggestion
| `policyContext.ip` | String | (Optional) The network rule condition, zone, or IP address. See [Network Condition](#network-condition-object) |
```

| `policyContext.zones.ids` | String | (Optional) The zone ID under the network rule condition. See [Network Condition](#network-condition-object) |
| `policyContext.risk` | String | (Optional) The risk rule condition: LOW, MEDIUM, HIGH. See [Risk Score Condition](#risk-score-condition-object)|
| `policyContext.device.managed` | Boolean | (Optional) If the device is registered. See [Device Condition](#device-condition-object)|
| `policyContext.device.registered` | Boolean | (Optional) If the device is managed. See [Device Condition](#device-condition-object)|

> | `policyContext.device.platform` |String | (Optional) The platform of the device, eg: IOS. See [Platform Condition](#platform-condition-object)|

Reviewer: ```suggestion
| `policyContext.device.platform` |String | (Optional) The platform of the device, for example: IOS. See [Platform Condition](#platform-condition-object)|
```

| `policyContext.device.assuranceId` | String | (Optional) The device assurance policy ID for the simulation|

##### Request example

> > **Note:** You can only evaluate `user` or `groups` not both, same for the `ip` and `zone.ids`. Use payload below for reference only

Reviewer: ```suggestion
> **Note:** You can only evaluate `user` or `groups` not both, and either `ip` or `zone.ids`, not both. Use the following payload for reference only.
```

```bash
curl -v -X POST \
-H "Accept: application/json" \
-H "Content-Type: application/json" \
-H "Authorization: SSWS ${api_token}" \
-d '
{
    "policyTypes": ["OKTA_SIGN_ON", "MFA_ENROLL"],
    "appInstance": "0oa4eroj3nYCIJIW70g7",
    "policyContext": {
        "user": {
            "id": "00u4eralyizp8NPOb0g7"
        },
        "groups": {
            "ids": [
              "00g4eralvekR5RLuS0g7", "00g4eralvekR5RLuS0g8"
            ]
        },
        "risk": {
            "level": "LOW"
        },
        "ip": "9.9.9.9",
        "zones": {
          "ids": [
            "nzo4eralxcRnbIHYJ0g7"
            ]
        },
        "device": {
            "platform": "IOS",
            "registered": true,
            "managed": true
        }
    }
}' "https://${yourOktaDomain}/api/v1/policies/simulate?expand=EVALUATED&expand=RULE"
```

#### Response

> The response section below explains the error responses and the response body

Reviewer: ```suggestion
The following response section explains the error responses and the response body.
```

> ##### Error Responses

Reviewer 2: Do we need to include error responses? Don't see it for other APIs on this page

Reviewer: I agree. I think we can leave these out.

HTTP 400:
Please provide the policy context for the request. ErrorCode: E0000001.

Please provide only group IDs or a user ID. ErrorCode: E0000001.

Please provide only zone IDs or an IP address. ErrorCode: E0000001.

This request contains an invalid policyType. ErrorCode: E0000001.

Please provide a valid application instance for the request. ErrorCode: E0000001.

> Please provide valid group IDs for the request. `groupIds` is an invalid ID. ErrorCode: E0000001.

Reviewer 2: `groupIds` -> `groupId`

Please provide a valid user for the request. ErrorCode: E0000001.

Access to `appName` is not allowed. `user or group` is not assigned to this application. ErrorCode: E0000001.

##### Response Body
| Property | Type                     | Description                          |
| -------- | -------------------------|--------------------------------------|
| `policyType` | String| The policy type we are simulating|
| `id` | String| ID of the specified policy/rule type|
| `name` | String| Policy name or policy rule name|

> | `status` | 	ENUM(MATCH, NOT_MATCH, UNDEFINED)| The result of this entity evaluation|

Reviewer: ```suggestion
| `status` | 	ENUM (MATCH, NOT_MATCH, UNDEFINED)| The result of this entity evaluation|
```
